lib/wpc_proto/internal_modules/proto_otap.c

```diff
@@ -30,27 +38,113 @@ void Proto_otap_close(void)
+                            total_size,
+                            seq);
+
+    /* Check if it is first chunk and no other load ongoing */
+    if (offset == 0)
+    {
+        if (m_scratchpad_load_current_seq != INVALID_CURRENT_SEQ)
+        {
+            LOGW("Previous upload not finshed: old seq %u, new seq %u\n",
+                                m_scratchpad_load_current_seq,
+                                seq);
+
+            /* Only a log, keep going with it */
+        }
+
+        if (WPC_get_stack_status(&status) == APP_RES_OK && status == 0)
+        {
+            /* Stack must be stoped */
+            /* No check required as next step will anyway fail */
+            WPC_stop_stack();
+            m_restart_after_load = true;
+        }
+
+        m_scratchpad_load_current_seq = seq;
+        res = WPC_start_local_scratchpad_update(total_size, seq);
+        if (res != APP_RES_OK)
+        {
+            return res;
+        }
+    }
+
+    /* Check chunk is from valid scratchpad */
+    if (seq != m_scratchpad_load_current_seq)
+    {
+        LOGE("Invalid scratcpad seq %u vs %u expected\n",
+                        seq,
+                        m_scratchpad_load_current_seq);
+        return APP_RES_INVALID_SEQ;
+    }
+
+    /* Load the chunk */
+    res = WPC_upload_local_block_scratchpad(chunk_size, chunk, offset);
+    if (res != APP_RES_OK)
+    {
+        return res;
+    }
+
+    /* Check if it is the last one */
+    if ((offset + chunk_size) == total_size)
+    {
+        /* Yes it is */
+        m_scratchpad_load_current_seq = INVALID_CURRENT_SEQ;
+        if (m_restart_after_load)
+        {
+            if (WPC_start_stack() != APP_RES_OK)
+            {
+                LOGE("Cannot restart stack after loading last chunk\n");
+            }
+            m_restart_after_load = false;
+        }
+    }
+
+    return APP_RES_OK;
+}
+
 app_proto_res_e Proto_otap_handle_upload_scratchpad(wp_UploadScratchpadReq *req,
                                                     wp_UploadScratchpadResp *resp)
 {
     app_res_e res = APP_RES_OK;
 
-    // TODO: Add some sanity checks
-
     if(req->has_scratchpad)
     {
 
-        /* Send the file to the sink */
-        res = WPC_upload_local_scratchpad(req->scratchpad.size,
-                                          req->scratchpad.bytes,
-                                          req->seq);
-        if (res == APP_RES_OK)
+        if (req->has_chunk_info)
         {
-            LOGI("Scratchpad uploaded : with seq %d of size %d\n", req->seq, req->scratchpad.size);
+            /* This scratchpad is a chunk */
+            res = handle_scratchpad_chunk(req->scratchpad.bytes,
+                                          req->scratchpad.size,
+                                          req->chunk_info.start_offset,
+                                          req->chunk_info.scratchpad_total_size,
+                                          req->seq);
```

Comment:
We could receive a 32bits seq, and we map it on 8bits. Should we check that range is ok to avoid strange behavior ? At least UINT32_MAX  is not allowed.

Reply:
Yes, I was also wondering where to put this limitation/check.